_posts/2020-11-2-kubernetes-webhooks-made-easy-with-openfaas-part-2.md

```diff
@@ -0,0 +1,291 @@
+
+* Deploy it project
+
+```sh
+$ cd deployment
+$ kubectl apply -f rbac.yaml,service.yaml,deployment.yaml
+# Label the default namespace to enable the webhook
+$ kubectl label namespaces default admission-webhook-example=enabled
+```
+
+* This time we are using "fileinjector" as a FUNCTION_NAME environment variable in the deployment.yaml file
+
+```yaml
+env:
+  - name: FUNCTION_NAME
+    value: fileinjector
+```
+
+### 5. Build and Deploy OpenFaaS Function (Optional)
+
+* Pull the [golang-middleware](https://github.com/openfaas-incubator/golang-http-template) template from [OpenFaaS Official Template Store](https://github.com/openfaas/store)
+
+```sh
+$ faas-cli template store list # check available templates in store
+$ faas-cli template store describe golang-middleware # describe the specific template
+$ faas-cli template store pull golang-middleware
+```
+
+* Create the function
+
+```sh
+$ export OPENFAAS_PREFIX=$DOCKER_USER
+$ faas-cli new fileinjector --lang go-middleware
+$ cd fileinjector
+$ go mod init fileinjector
+$ # fill the handler.go with the corresponding code: [functions/fileinjector/handler.go](https://github.com/developer-guy/admission-webhook-example-with-openfaas/blob/feature/mutating/functions/fileinjector/handler.go)
+$ go get
+```
+
+* Deploy the function
+
+```sh
+$ cd functions
+$ faas-cli up -f fileinjector.yml --build-arg GO111MODULE=on # (build-push-deploy) make sure you are using your docker hub username. i.e: devopps
+```
+
+* Verify the functions that are working in `openfaas-fn` namespace
+
+```sh
+$ kubectl get pods --namespace openfaas-fn
+```
+
+### 6. Testing the whole workflow
+* The purpose of this PoC is that to add a simple volume which is type configMap and volumeMount which is mounted to that volume to the pod automatically by our Mutating Admission Webhook. 
+
+* First, we need to create a ConfigMap to be able to mount a file to the Pod:
+
+```sh
+$ cat functions/fileinjector/hello-openfaas.txt
+
+  _  _     _ _        ___                 ___          ___
+ | || |___| | |___   / _ \ _ __  ___ _ _ | __|_ _ __ _/ __|
+ | __ / -_) | / _ \ | (_) | '_ \/ -_) ' \| _/ _` / _` \__ \
+ |_||_\___|_|_\___/  \___/| .__/\___|_||_|_|\__,_\__,_|___/
+                          |_|
+```
+
+```
+$ kubectl create hello-configmap --from-file=functions/fileinjector/hello-openfaas.txt
+```
+
+* Now , you can apply the Pod manifest, once you done, you will notice that Pod includes a volume + volumeMount additional to the manifest you wrote in the beginning
+
+```sh
+$ kubectl apply -f busybox.yaml
+```
+
+* Check the logs of the Pod, you will see "Hello OpenFaaS" text as the output of the container.
+
+```sh
+$ kubectl logs -f busybox
+
+  _  _     _ _        ___                 ___          ___
+ | || |___| | |___   / _ \ _ __  ___ _ _ | __|_ _ __ _/ __|
+ | __ / -_) | / _ \ | (_) | '_ \/ -_) ' \| _/ _` / _` \__ \
+ |_||_\___|_|_\___/  \___/| .__/\___|_||_|_|\__,_\__,_|___/
+                          |_|
+```
+
```

Comment:
You are missing the "What-if" section of the 4MAT system I asked you to follow before I reviewed the post.

What - what is this thing? And what is it not?
Why - why did you build it? Why does it need to exist?
How - how do I deploy it? What examples do you have and use-cases?
What if - what can I go on to do with this? Can I extend it, can I contribute? Can I apply it to a problem at work? Is it suitable for production?

Reply:
I added a new section called "What can you do with it ?"
Done.